### PR TITLE
Add Lombok-style @Singular support for classic builders

### LINKED ISSUE #48: Support `@Singular`-like per-element collection builder methods
### Motivation
Lombok's `@Singular` generates single-item and bulk-item adder methods for collection fields, greatly improving builder ergonomics.  
Jilt already interoperates with Lombok but lacks this capability, forcing callers to pre-assemble collections before using the builder.

### Desired behaviour
For any builder input  (field/parameter) marked `@Singular`:

* Generate  
  * `<B addElem(E e)>` – adds one element  
  * `<B addAllElems(Collection<? extends E> c)>` – adds many  
* Generated builder accumulates elements internally (e.g. `List<E> _elems = new ArrayList<>()`) and converts to the declared collection type in `build()`.  
* Works in both classic and staged builders, preserving fluent typing.
* Initial support for Lists. Sets and Maps can be added in a later release.

### Example

```java
@Singular           // new annotation in org.jilt
private List<String> tags;
```
produces
```java
public Builder addTag(String tag) { /* ... */ }
public Builder addAllTags(Collection<? extends String> tags) { /* ... */ }
```

Thanks for considering!

### PR DESCRIPTION
**Pull request by Chazmus**
Implements Lombok-style @Singular support for classic builders, generating adder methods for List fields.

Details:
- Adds @Singular annotation.
- Classic builders generate addX and addAllX methods for singular fields.
- Accumulators are assigned directly in build().
- Staged builders ignore @Singular.

Tests:
Includes tests for singular builder ergonomics.

Fixes #48

### CODE REVIEW

**Comment by Chazmus** [comment]
Full disclosure: This is AI slop... but looks alright from my personal review.

File: src/test/java/org/jilt/test/SingularDemoTest.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import static org.junit.Assert.*;
+
+public class SingularDemoTest {
+    @Test
+    public void testSingularAddMethods() {
+        SingularDemoBuilder builder = SingularDemoBuilder.singularDemo();
+        builder.addTag("foo");
+        builder.addTag("bar");
+        builder.addAllTags(Arrays.asList("baz", "qux"));
+        SingularDemo demo = builder.build();
```
Comment (skinny85) [inline src/test/java/org/jilt/test/SingularDemoTest.java:16]: Can you use a fluent style instead?

```suggestion
        SingularDemo demo = SingularDemoBuilder.singularDemo()
            .addTag("foo");
            .addTag("bar");
            .addAllTags(Arrays.asList("baz", "qux"))
            .build();
```

File: src/test/java/org/jilt/test/SingularDemoTest.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        List<String> tags = demo.getTags();
+        assertEquals(Arrays.asList("foo", "bar", "baz", "qux"), tags);
+        assertTrue(tags instanceof java.util.ArrayList);
```
Comment (skinny85) [inline src/test/java/org/jilt/test/SingularDemoTest.java:19]: Please use AssertJ instead. There are examples everywhere in the repo.

File: src/test/java/org/jilt/test/SingularDemoTest.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    @Test
+    public void testClassicSetterIgnoredForSingular() {
+        SingularDemoBuilder builder = SingularDemoBuilder.singularDemo();
+        // builder.tags(Arrays.asList("should", "not", "exist")); // Should not compile
+        builder.addTag("only");
+        SingularDemo demo = builder.build();
+        assertEquals(Arrays.asList("only"), demo.getTags());
```
Comment (skinny85) [inline src/test/java/org/jilt/test/SingularDemoTest.java:28]: Same comments here (fluent style, AssertJ).

File: src/main/java/org/jilt/internal/AbstractBuilderGenerator.java [hunk 4]
```diff
@@ -95,6 +95,52 @@ public final void generateBuilderClass() throws Exception {
...
+
+                    builderClassBuilder.addMethod(MethodSpec.methodBuilder(addAllName)
+                        .addModifiers(Modifier.PUBLIC)
+                        .returns(this.builderClassTypeName())
+                        .addParameter(ParameterizedTypeName.get(ClassName.get("java.util", "Collection"), WildcardTypeName.subtypeOf(pType.typeArguments.get(0))), fieldName)
+                        .addStatement("this.$L.addAll($L)", accName, fieldName)
+                        .addStatement("return this")
+                        .build());
+
+                    // Do not generate normal setter for singular fields in classic builder
```
Comment (skinny85) [inline src/main/java/org/jilt/internal/AbstractBuilderGenerator.java:139]: I think this is a mistake We should still generate a "regular" setter too. Otherwise, there will be no way to "reset" the elements already added to the collection.

File: src/main/java/org/jilt/internal/AbstractBuilderGenerator.java [hunk 5]
```diff
@@ -95,6 +95,52 @@ public final void generateBuilderClass() throws Exception {
...
+                        .returns(this.builderClassTypeName())
+                        .addParameter(pType.typeArguments.get(0), singular)
+                        .addStatement("this.$L.add($L)", accName, singular)
+                        .addStatement("return this")
+                        .build());
+
+                    builderClassBuilder.addMethod(MethodSpec.methodBuilder(addAllName)
+                        .addModifiers(Modifier.PUBLIC)
+                        .returns(this.builderClassTypeName())
+                        .addParameter(ParameterizedTypeName.get(ClassName.get("java.util", "Collection"), WildcardTypeName.subtypeOf(pType.typeArguments.get(0))), fieldName)
```
Comment (skinny85) [inline src/main/java/org/jilt/internal/AbstractBuilderGenerator.java:134]: What if the type was already a wildcard? Something like `List<? extends String>`? Can you please add a test covering that corner case?

**Review by skinny85 — COMMENTED** [review]
Thanks, I hate it 😛.

A few comments, let me know if you agree.

File: src/main/java/org/jilt/Singular.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+import java.lang.annotation.*;
+
+/**
+ * Marks a collection field or parameter for builder singular methods.
+ * Only supported for List types in initial release.
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.FIELD, ElementType.PARAMETER})
+public @interface Singular {
```
[the remark this answers is not in the document]
Comment (Chazmus) [reply src/main/java/org/jilt/Singular.java:11]: The org at which I work doesn't want to use Lombok (some divisiveness regarding the way Lombok does some funky class manipulation). But we're keen to  introduce Jilt and refactor our builders. So I'd somewhat push back on this as it would require the application to have deps on both Jilt and Lombok... How about we do both? Have an interface in Jilt OR allow you to use the Lombok one?

**Comment by skinny85** [comment]
BTW, can you show me a compelling example of how this feature makes code look significantly better? Maybe with an example from your codebase (appropriately annonomyzed, of course)?

File: src/main/java/org/jilt/internal/AbstractBuilderGenerator.java [hunk 7]
```diff
@@ -199,12 +245,32 @@ private MethodSpec makeBuildMethod() {
...
+                    }
+                }
+                args.add(fieldName);
+            }
+            String attributes = String.join(", ", args);
             if (this.targetCreationMethodIsConstructor()) {
                 buildMethod.addStatement("return new $T($L)", this.targetClassTypeName(), attributes);
             } else {
                 buildMethod.addStatement("return $T.$L($L)",
-                        // using ClassName gets rid of any type parameters the class might have
```
[the remark this answers is not in the document]
Comment (Chazmus) [reply src/main/java/org/jilt/internal/AbstractBuilderGenerator.java:207]: It wasn't me! It was the robot :robot: (I'll put it back)

**Comment by Chazmus** [comment]
> BTW, can you show me a compelling example of how this feature makes code look significantly better? Maybe with an example from your codebase (appropriately annonymyzed, of course)?

So, in our code base we have several instances where we'd create a builder, then pass the builder around to a bunch of different places/classes where the code will choose to set things on the builder or add to lists depending on various criteria, ultimately returning to it's original place to be .build()'d. I know this could be overcome by also say, passing a list around or by subclassing the generated builder (in the case where we'd want to use Jilt) but that would add back to boilerplate that we're essentially trying to save on.

If you're not keen on adding this feature in the interest of keeping Jilt lean, I'm totally happy to drop working on this! Though it does seem like I'm not the only one who thinks this could be useful judging from the open issue, and the fact that this exists in Lombok.

Let me know your thoughts! (And sorry for the AI pass, after having a proper look at it and reading through your comments, it was a bit rubbish :laughing: ) I've refactored it in the way that I believe you were implying in your comments, let me know if I'm still going down the wrong path. Thanks in advance!

**Comment by skinny85** [comment]
I'm wondering: would adding the option to the Builder to generate getters for its properties would actually be a better solution for the problem you've described? Then, when passing the Builder to those other methods, they can make any modifications to the `List` that is stored in a given property that they want, not limited to just the `addTag()` / `andAllTags()` methods that we have to add ourselves to the Builder (for example, why not `clearTag()`? Why not `addAllTags()` that takes in a `Stream`? Etc.). Instead, when they get the underlying list of tags with `getTags()`, they can modify it in any way that they need.

Thoughts on this idea?

**Comment by skinny85** [comment]
> * Would you like the new getters to be hidden behind a config flag/annotation?

Yes. They should not be generated by default, but only if a new attribute of the `@Builder` annotation, `generateGetters`, of type `boolean` (default value: `false`), is set to `true`.

> * Are you happy for the getters to be on all types of builders (just putting the code in the `AbstractBuilderGenerator`?)

Almost. They should only be present on the Builder class (not added to any of the Staged Builder interfaces). _However_, for the `FUNCTIONAL` `BuilderStyle`, this attribute should either be ignored, or cause a compilation error, but it should _not_ generate any getters in that case (I can explain why, if you want me to expand on the justification).

> * Any other specific implementation details you can think of that might be nice to have/worth considering?

Avoid code like `this.getClass().getSimpleName().equals("ClassicBuilderGenerator")`, and we should be good 😉.

**Comment by skinny85** [comment]
@Chazmus are you still working on this PR?

**Comment by Chazmus** [comment]
> @Chazmus are you still working on this PR?

Hi @skinny85, I have been meaning too but been quite busy recently. I'll close this for now and raise a new one once done. Apologies!
